File: public/cpp/049-alien-dictionary.cpp
/*
 * LeetCode Problem #269: Alien Dictionary
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/alien-dictionary/
 *
 * Problem Statement:
 * There is a new alien language that uses the English alphabet. The order of letters
 * is unknown to you. You are given a list of strings words from the alien language's
 * dictionary, where the strings in words are sorted lexicographically by the rules
 * of this new language.
 *
 * Return a string of the unique letters in the new language sorted by the rules,
 * or "" if the order cannot be determined, or "" if there is a contradiction.
 *
 * Example 1:
 *   Input: words = ["wrt","wrf","er","ett","rftt"]
 *   Output: "wertf"
 *
 * Example 2:
 *   Input: words = ["z","x"]
 *   Output: "zx"
 *
 * Example 3:
 *   Input: words = ["z","x","z"]
 *   Output: "" (contradiction — cycle)
 *
 * Constraints:
 *   - 1 <= words.length <= 100
 *   - 1 <= words[i].length <= 100
 *   - words[i] consists of lowercase English letters
 */

#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <iostream>
using namespace std;

// ==================== APPROACH 1: Topological Sort via BFS (Kahn's Algorithm) ====================
// 1. Build directed graph from adjacent word comparisons.
// 2. Run Kahn's BFS topological sort.
// Time: O(C) where C = total characters across all words  Space: O(1) (26 chars at most)
class Solution1 {
public:
    string alienOrder(vector<string>& words) {
        // Collect all unique characters
        unordered_map<char,int> indegree;
        unordered_map<char,unordered_set<char>> adj;
        for (auto& w : words) for (char c : w) if (!indegree.count(c)) indegree[c] = 0;

        // Build edges
        for (int i = 0; i+1 < (int)words.size(); i++) {
            string& a = words[i], &b = words[i+1];
            int len = min(a.size(), b.size());
            // Prefix longer word = invalid
            if (a.size() > b.size() && a.substr(0, len) == b.substr(0, len)) return "";
            for (int j = 0; j < (int)len; j++) {
                if (a[j] != b[j]) {
                    if (!adj[a[j]].count(b[j])) { adj[a[j]].insert(b[j]); indegree[b[j]]++; }
                    break;
                }
            }
        }

        // BFS topological sort
        queue<char> q;
        for (auto& [c, d] : indegree) if (d == 0) q.push(c);
        string result;
        while (!q.empty()) {
            char c = q.front(); q.pop();
            result += c;
            for (char nb : adj[c]) if (--indegree[nb] == 0) q.push(nb);
        }
        return result.size() == indegree.size() ? result : "";
    }
};

// ==================== APPROACH 2: Topological Sort via DFS (Cycle Detection) ====================
// DFS post-order gives reverse topological order; detect cycles with coloring.
// Time: O(C)  Space: O(1)
class Solution2 {
    unordered_map<char, vector<char>> adj;
    unordered_map<char, int> color; // 0=unvisited,1=visiting,2=done
    string order;
    bool hasCycle;

    void dfs(char c) {
        if (hasCycle) return;
        color[c] = 1;
        for (char nb : adj[c]) {
            if (color[nb] == 1) { hasCycle = true; return; }
            if (color[nb] == 0) dfs(nb);
        }
        color[c] = 2;
        order += c;
    }
public:
    string alienOrder(vector<string>& words) {
        unordered_set<char> chars;
        for (auto& w : words) for (char c : w) { chars.insert(c); color[c] = 0; }

        for (int i = 0; i+1 < (int)words.size(); i++) {
            string& a = words[i], &b = words[i+1];
            int len = min(a.size(), b.size());
            if (a.size() > b.size() && a.substr(0,len) == b.substr(0,len)) return "";
            for (int j = 0; j < (int)len; j++) {
                if (a[j] != b[j]) { adj[a[j]].push_back(b[j]); break; }
            }
        }

        hasCycle = false;
        for (char c : chars) if (color[c] == 0) dfs(c);
        if (hasCycle) return "";
        reverse(order.begin(), order.end());
        return order;
    }
};

// ==================== APPROACH 3: BFS with explicit set (cleaner) ====================
// Same as Approach 1 but uses set instead of unordered_set for deterministic order.
class Solution3 {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char,int> indegree;
        unordered_map<char,vector<char>> adj;
        for (auto& w : words) for (char c : w) if (!indegree.count(c)) indegree[c] = 0;
        for (int i = 0; i+1 < (int)words.size(); i++) {
            auto& a = words[i]; auto& b = words[i+1];
            int len = min(a.size(), b.size());
            if (a.size() > b.size() && a.substr(0,len)==b.substr(0,len)) return "";
            for (int j = 0; j < (int)len; j++) {
                if (a[j] != b[j]) { adj[a[j]].push_back(b[j]); indegree[b[j]]++; break; }
            }
        }
        queue<char> q;
        for (auto& [c,d] : indegree) if (d==0) q.push(c);
        string res;
        while (!q.empty()) {
            char c = q.front(); q.pop(); res += c;
            for (char nb : adj[c]) if (--indegree[nb]==0) q.push(nb);
        }
        return res.size()==indegree.size() ? res : "";
    }
};

void runTests() {
    Solution1 sol;
    vector<string> w1 = {"wrt","wrf","er","ett","rftt"};
    cout << "Test 1: \"" << sol.alienOrder(w1) << "\" (expected \"wertf\")\n";
    vector<string> w2 = {"z","x"};
    cout << "Test 2: \"" << sol.alienOrder(w2) << "\" (expected \"zx\")\n";
    vector<string> w3 = {"z","x","z"};
    cout << "Test 3: \"" << sol.alienOrder(w3) << "\" (expected \"\")\n";
}

int main() { runTests(); return 0; }

// Made with Bob

File: public/python/049-alien-dictionary.py
# LeetCode Problem #269: Alien Dictionary
# Difficulty: Hard
# Link: https://leetcode.com/problems/alien-dictionary/

from typing import List
from collections import deque, defaultdict


# ─────────────────────────────────────────────
# APPROACH 1: Topological Sort via BFS (Kahn's Algorithm) | O(C) time | O(1) space
# EXPLAIN: Extract ordering constraints from adjacent word pairs; run Kahn's BFS topological sort; cycle ↔ return "".
# WHEN: Standard interview answer; clean iterative implementation.

def alienOrder_bfs(words: List[str]) -> str:
    indegree = {c: 0 for w in words for c in w}
    adj: dict = {c: [] for c in indegree}
    for i in range(len(words) - 1):
        a, b = words[i], words[i+1]
        mn = min(len(a), len(b))
        if len(a) > len(b) and a[:mn] == b[:mn]:
            return ""
        for j in range(mn):
            if a[j] != b[j]:
                adj[a[j]].append(b[j])
                indegree[b[j]] += 1
                break
    q = deque(c for c in indegree if indegree[c] == 0)
    result = []
    while q:
        c = q.popleft()
        result.append(c)
        for nb in adj[c]:
            indegree[nb] -= 1
            if indegree[nb] == 0:
                q.append(nb)
    return ''.join(result) if len(result) == len(indegree) else ""


# ─────────────────────────────────────────────
# APPROACH 2: Topological Sort via DFS (Cycle Detection) | O(C) time | O(1) space
# EXPLAIN: DFS post-order gives reverse topological order; detect cycles with 3-color marking.
# WHEN: Recursive approach; post-order naturally produces the reverse order.

def alienOrder_dfs(words: List[str]) -> str:
    adj: dict = defaultdict(list)
    chars = {c for w in words for c in w}
    color = {c: 0 for c in chars}  # 0=unvisited, 1=visiting, 2=done
    order = []
    has_cycle = [False]

    for i in range(len(words) - 1):
        a, b = words[i], words[i+1]
        mn = min(len(a), len(b))
        if len(a) > len(b) and a[:mn] == b[:mn]:
            return ""
        for j in range(mn):
            if a[j] != b[j]:
                adj[a[j]].append(b[j])
                break

    def dfs(c):
        if has_cycle[0]: return
        color[c] = 1
        for nb in adj[c]:
            if color[nb] == 1:
                has_cycle[0] = True
                return
            if color[nb] == 0:
                dfs(nb)
        color[c] = 2
        order.append(c)

    for c in chars:
        if color[c] == 0:
            dfs(c)
    if has_cycle[0]:
        return ""
    return ''.join(reversed(order))


# ─────────────────────────────────────────────
# APPROACH 3: BFS with Explicit Set (Cleaner Variant) | O(C) time | O(1) space
# EXPLAIN: Same as Approach 1 but uses a list for adj values instead of set, deduplicating with a seen-edges set.
# WHEN: Cleaner deduplication; functionally equivalent to Approach 1.

def alienOrder_bfs_clean(words: List[str]) -> str:
    indegree = {c: 0 for w in words for c in w}
    adj: dict = defaultdict(list)
    seen_edges: set = set()
    for i in range(len(words) - 1):
        a, b = words[i], words[i+1]
        mn = min(len(a), len(b))
        if len(a) > len(b) and a[:mn] == b[:mn]:
            return ""
        for j in range(mn):
            if a[j] != b[j]:
                if (a[j], b[j]) not in seen_edges:
                    seen_edges.add((a[j], b[j]))
                    adj[a[j]].append(b[j])
                    indegree[b[j]] += 1
                break
    q = deque(c for c in indegree if indegree[c] == 0)
    result = []
    while q:
        c = q.popleft()
        result.append(c)
        for nb in adj[c]:
            indegree[nb] -= 1
            if indegree[nb] == 0:
                q.append(nb)
    return ''.join(result) if len(result) == len(indegree) else ""


# Made with Bob
